src/Plutarch/Extra/AssetClass.hs

```diff
@@ -1,97 +1,361 @@
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE UndecidableInstances #-}
 
+{- | Provides Data and Scott encoded  asset class types and utility
+  functions.
+-}
 module Plutarch.Extra.AssetClass (
-    PAssetClass (..),
-    passetClass,
-    passetClassValueOf,
-    pvalueOf,
+    -- * AssetClass - Hask
+    AssetClass (AssetClass, symbol, name),
+    assetClassValue,
+
+    -- * AssetClass - Plutarch
+    PAssetClass (PAssetClass, psymbol, pname),
+    adaSymbolData,
+    isAdaClass,
+    adaClass,
+    padaClass,
+    emptyTokenNameData,
+    pcoerceCls,
+    psymbolAssetClass,
+    pconstantCls,
+
+    -- * AssetClassData - Hask
+    AssetClassData (AssetClassData, symbol, name),
+
+    -- * AssetClassData - Plutarch
+    PAssetClassData (PAssetClassData),
+
+    -- * Scott <-> Data conversions
+    toScottEncoding,
+    ptoScottEncoding,
+    fromScottEncoding,
+    pfromScottEncoding,
+    pviaScottEncoding,
 ) where
 
+--------------------------------------------------------------------------------
+
+import GHC.TypeLits (Symbol)
+
+--------------------------------------------------------------------------------
+
+import qualified Data.Aeson as Aeson
+import Data.Tagged (Tagged, untag)
+import qualified Generics.SOP as SOP
+import qualified PlutusLedgerApi.V1.Value as Value
+
+--------------------------------------------------------------------------------
+
+import Plutarch.DataRepr (PDataFields)
+
+import Plutarch.Extra.IsData (
+    DerivePConstantViaDataList (DerivePConstantViaDataList),
+    ProductIsData (ProductIsData),
+ )
+import Plutarch.Lift (
+    PConstantDecl,
+    PUnsafeLiftDecl (PLifted),
+ )
+
+import qualified PlutusTx
+
+--------------------------------------------------------------------------------
+
+import Plutarch.Extra.Record (mkRecordConstr, (.&), (.=))
+
+import Data.Aeson (FromJSON, FromJSONKey, ToJSON, ToJSONKey)
+import qualified GHC.Generics as GHC
+
+--------------------------------------------------------------------------------
+
 import Plutarch.Api.V1 (
     PCurrencySymbol,
     PTokenName,
-    PValue (PValue),
- )
-import Plutarch.Api.V1.AssocMap (PMap (PMap))
-import Plutarch.Api.V2 (
-    AmountGuarantees,
-    KeyGuarantees,
  )
-import Plutarch.DataRepr (PDataFields)
-import Plutarch.Extra.TermCont (pletC, pmatchC)
+import Plutarch.Unsafe (punsafeCoerce)
+
+--------------------------------------------------------------------------------
+
+import Plutarch.Orphans ()
+
+--------------------------------------------------------------------------------
+
+import qualified PlutusLedgerApi.V1.Value as Plutus
+
+--------------------------------------------------------------------------------
+-- AssetClass & Variants
+
+{- | Version of Plutus AssetClass, with currency unit tag
+ @since 3.8.0
+-}
+data AssetClass (tag :: Symbol) = AssetClass
+    { symbol :: Plutus.CurrencySymbol
+    , name :: Plutus.TokenName
```

Comment:
Need `since`s on these, since they only appeared under their current names in this version. Also, would `CurrencySymbol` and `TokenName` conflict with something? If so, they should be imported unqualified.

Reply:
Good call. These were qualified downstream, but _if_ there was a conflict, there isn't any more.